### DISCUSSION TITLE
DJL in Robocode

### DISCUSSION THREAD

**Discussion by aurora126** [deepjavalibrary/djl#1848] [Q&A] [answered]
Hi All,

First of all this is a great API.

Next off I'm trying to build an RL robot in [robocode](https://robocode.sourceforge.io/) using DJL but getting a bit stuck because of robocode's security manager which prevents the Engine reading the environment variable DJL_DEFAULT_ENGINE, I think also it may be prevented accessing the java property ai.djl.default_engine as well.

Preventing test.BaseRLRobot* from access: ("java.lang.RuntimePermission" "getenv.DJL_DEFAULT_ENGINE")
.test.BaseRLRobot* has caused a security violation. This robot has been banned and will not be allowed to participate in battles.
test.BaseRLRobot*: Throwable: java.lang.ExceptionInInitializerError

I can't really modify robocode as it means I would not be able to compete in the competitions.

Is there any other way around this?  

Thanks

**Comment by frankfliu** [comment]
1. You can always programmatically set the DJL' system properties : `System.setProperties("xxxx", "")`
2. If you only include one engine in your classpath, the engine will be the default engine
3. You can choose which engine to use at runtime:
```
Engine engine = Engine.getEngine("PyTorch");
NDManager manager = engine.newBaseManager();

OR:
NDManager manager = NDManager.newBaseManager(Device.gpu(0), "PyTorch");

For load the model:

        Criteria<Image, DetectedObjects> criteria =  Criteria.builder()
                        .setTypes(Image.class, DetectedObjects.class)
                        .optEngine("PyTorch")

```

**Reply by aurora126** [reply]
Hi @frankfliu 

Many thanks for the reply. I've tried both of the options you suggest above but both of them seem to issue a call to get the environment variable DJL_DEFAULT_ENGINE.   I only have the PyTorch engine in my classpath.

I think the problem I have is that in the Engine class it has the declaration 

private static final String DEFAULT_ENGINE = initEngine();

and in the initEngine method it calls 

String defaultEngine = System.getenv("DJL_DEFAULT_ENGINE");

and this is being blocked by the robocode security manager.

Any other way around this at all?

<html>
<body>

SYSTEM: Preventing test.BaseRLRobot* from access: ("java.lang.RuntimePermission" "getenv.DJL_DEFAULT_ENGINE")
java.lang.ExceptionInInitializerError
at ai.djl.ndarray.NDManager.newBaseManager(NDManager.java:136)
at test.BaseRLRobot.initialiseRobot(BaseRLRobot.java:59)
at test.BaseRLRobot.run(BaseRLRobot.java:43)
at net.sf.robocode.host.proxies.HostingRobotProxy.callUserCode(HostingRobotProxy.java:290)
at net.sf.robocode.host.proxies.HostingRobotProxy.run(HostingRobotProxy.java:229)
at net.sf.robocode.host.proxies.BasicRobotProxy.run(BasicRobotProxy.java:42)
at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "getenv.DJL_DEFAULT_ENGINE")
at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:485)
at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
at net.sf.robocode.host.security.RobocodeSecurityManager.checkPermission(RobocodeSecurityManager.java:138)
at java.base/java.lang.System.getenv(System.java:1118)
at ai.djl.engine.Engine.initEngine(Engine.java:74)
at ai.djl.engine.Engine.(Engine.java:56)

</body>
</html>

**Reply by frankfliu** [reply]
OK, I see your problem.

We are trying to access environment variable, and it throws permission error. One thing we can do is try to access the environment variable, and ignore the error. I will take a look and see If can can create a PR

**Reply by aurora126** [reply]
Many thanks @frankfliu that would be great.

**Comment by frankfliu** [comment]
@aurora126 
https://github.com/deepjavalibrary/djl/pull/1852 PR fixes your issue. You can try our SNAPSHOT build

**Reply by aurora126** [reply]
Hi @frankfliu 

Many thanks for implementing this.

I probably should have checked whether what you implemented would work or not first.

Robocode now produces the following ...
Preventing test.BaseRLRobot* from access: ("java.lang.RuntimePermission" "getenv.DJL_DEFAULT_ENGINE")

It looks like the Java Security Manager has been deprecated in Java 18.   In the latest version of Robocode in order to support Java 18 (where it is deprecated) they have temporarily (I assume) switched it back on 
(https://github.com/robo-code/robocode/blob/master/versions.md) using the JVM parameter -Djava.security.manager=allow.

I tried creating some code to check the runtime permission 
`java.lang.RuntimePermission permission = new java.lang.RuntimePermission("getenv.{variable name}");
		SecurityManager appsm = System.getSecurityManager();
		try
		{
			appsm.checkPermission(permission);
			System.out.println("Permission ok getenv.{variable name}");
		}
		catch(SecurityException ex)
		{
			System.out.println("SecurityException no security permission to access getenv.{variable name}" + ex.toString());
		}`

but that also seemed to get detected by the robocode security manager (I'm not quite sure how).

It looks like I can turn off the robocode security manager by passing it -DNOSECURITY=true.

So for now I would suggest not doing anymore and I'll make some enquiries with the robocode team.

Once again many thanks for your help.

**Comment by aurora126** [comment, marked as answer]
Just an update on this thread for anyone interested.  There is a newer version of robocode in development ( [Robocode Tank Royale](https://robocode-dev.github.io/tank-royale/) ) , each robot runs in its own JVM and there are no security restrictions.  At the moment there are no competitions for the newer version but I assume there will be in the near future.

For now the robocode team have fixed the issue with current version ([Robocode](https://robocode.sourceforge.io/)) to enable access to System.getenv.

Hope this helps anyone wanting to use DJL for build a robocode robot.
